### PR TITLE
Add log_queries_probability setting

### LINKED ISSUE #16609: Support logging a random sample of queries to query_log
With a large volume of queries per second logging all queries gets expensive, it would be nice to be able to log only a configurable fraction of the queries, selected at random.

### PR DESCRIPTION
**Pull request by evillique**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add `log_queries_probability` setting that allows user to write to query_log only a sample of queries. Closes #16609.

### CODE REVIEW

**Comment by amosbird** [comment]
Can we make this random process consistent across a cluster?

**Comment by akuzm** [comment]
> Can we make this random process consistent across a cluster?

By the way, this is one of things solved by distributed tracing protocols such as OpenTelemetry. If you choose to trace the top-level request, all downstream requests are also traced ("head-based sampling").

**Comment by Algunenano** [comment]
Does it make sense for this setting to cover other query-related logs too (thread_log)?

**Comment by alexey-milovidov** [comment]
> Does it make sense for this setting to cover other query-related logs too (thread_log)?

Yes, it should.

**Comment by alexey-milovidov** [comment]
> Can we make this random process consistent across a cluster?

Yes, it is possible.
We can modify `log_queries_probability` and set it either to 0 or 1 before sending internal request to remote servers ("probability collapse"). See MultiplexedConnections.cpp and HedgedConnections.cpp

File: src/Interpreters/executeQuery.cpp [hunk 1]
```diff
@@ -651,7 +653,10 @@ static std::tuple<ASTPtr, BlockIO> executeQueryImpl(
 
             elem.client_info = client_info;
 
-            bool log_queries = settings.log_queries && !internal;
+            std::bernoulli_distribution should_log{
```
Comment (alexey-milovidov) [inline src/Interpreters/executeQuery.cpp:656]: You've made it only for query start.
There is also query finish event.

And it needs to be consistent: if query is logged - it applied both for start and finish.

**Review by alexey-milovidov — CHANGES_REQUESTED** [review]
.

File: src/Interpreters/executeQuery.cpp (hunk 1 above)
Comment (evillique) [reply src/Interpreters/executeQuery.cpp:656]: Is not this exact bool variable `log_queries` captured in finish callback (line 730) and then used to determine whether to write query to log (line 797)?

File: src/Client/MultiplexedConnections.cpp [hunk 2]
```diff
@@ -119,6 +119,12 @@ void MultiplexedConnections::sendQuery(
 
     Settings modified_settings = settings;
 
+    /// Consistent write to log_queries across replicas
```
Comment (alexey-milovidov) [inline src/Client/MultiplexedConnections.cpp:122]: It needs to be the same as for main query.
So, if the main query is logged, the remote queries should also.

**Review by alexey-milovidov — COMMENTED** [review]
Now it's perfect!
But something wrong with integration tests...

**Review by alexey-milovidov — CHANGES_REQUESTED** [review]

**Review by alexey-milovidov — APPROVED** [review]

**Review by alexey-milovidov — CHANGES_REQUESTED** [review]
It is possible to provide a test.

1. Set probability to 0.5.
Run query 100 times.
And check that it sometimes logged and sometimes not.

2. Do the same for query FROM remote(test_shard_...)
Check that when it is logged, the subqueries are also logged and vice versa.

**Review by alexey-milovidov — APPROVED** [review]
Ok.

PS. It was also possible to do it inside functional test.

**Comment by gyuton** [comment]
Internal documentation ticket: DOCSUP-14537.

**Comment by Frank030366** [comment]
Hi folks - sorry for posting here, but it seems that there is a bug related to saving this settings through profile or interactively in client through SET command. Test case is simple (reproducing even in the latest versions):

- write this setting in default (for example) profile, placing the following XML in users.d dir:
<clickhouse>
<log_queries_probability>0.1</log_queries_probability>
</clickhouse>

- we can even restart clickhouse instance although this setting should be read online. After that check it in clickhouse-client:

:) show settings like 'log_queries_probability'

```
SHOW SETTINGS LIKE 'log_queries_probability'

Query id: 07f869ac-338c-41a3-8714-8f50c76822eb

┌─name────────────────────┬─type──┬─value─┐
│ log_queries_probability │ Float │ 1     │
└─────────────────────────┴───────┴───────┘

```
We see, that it's 1 regardless of what we've set in the profile. Ok, let's try to set it manually and check:

```
 :) set log_queries_probability=0.5

SET log_queries_probability = 0.5

Query id: 014ae9ff-912e-4bc7-826d-118df1ee02bd

Ok.

0 rows in set. Elapsed: 0.001 sec.

 :) show settings like 'log_queries_probability'

SHOW SETTINGS LIKE 'log_queries_probability'

Query id: 0d5920b7-229c-43d4-9081-afc2dfa4cfb3

┌─name────────────────────┬─type──┬─value─┐
│ log_queries_probability │ Float │ 1     │
└─────────────────────────┴───────┴───────┘
```

No any effect, it's still 1. But in addition we can see another interesting thing - let's try to set it to wrong value (>1):

```
 :) set log_queries_probability=1.5

SET log_queries_probability = 1.5

Query id: a3f8d632-e4b8-49c4-ae45-d294eb92001f

Ok.

0 rows in set. Elapsed: 0.001 sec.

 :) show settings like 'log_queries_probability'

SHOW SETTINGS LIKE 'log_queries_probability'

Query id: 2ff27b82-1a0a-4197-a3fa-cdc709f9b06b

┌─name────────────────────┬─type──┬─value─┐
│ log_queries_probability │ Float │ 1.5   │
└─────────────────────────┴───────┴───────┘
```

That's it - so we have a bug definitely.

**Comment by Algunenano** [comment]
> That's it - so we have a bug definitely.

It's not "definitely" a bug. It's always showing 1 because CH sets the value once to get consistency in distributed systems:

https://github.com/ClickHouse/ClickHouse/blob/ff1c0023e8ac614bec9df4d26251f7accca477b2/src/Interpreters/executeQuery.cpp#L909-L915

Basicaly if it's less than 1 it sets it to `log_queries_probability` and it sets `log_queries` to either 0 or 1 depending on the probability. The side effect is that you see these values in the queries.

**Pull request merged**